students/student_framllat/session04/mailroom2.py

```diff
@@ -0,0 +1,247 @@
+    """
+    Execute the logic to record a donation and generate a thank you message.
+    """
+    # Read a valid donor to send a thank you from, handling special commands to
+    # let the user navigate as defined.
+    while True:
+        name = input("Enter a donor's name or 'list' to see all donors or "
+                     "'menu' to exit to main menu > ").strip()
+        if name == "list":
+            print(list_donors())
+        elif name == "menu":
+            return
+        else:
+            take_donation(name)
+
+
+def sort_key(item):
+    # used to sort on name in donor_db
+    # NOTE: this could be replaced by ``itemgetter`` in the
+    #       ``operator`` module
+    return item[1]
+
+
+def generate_donor_report():
+    """
+    Generate the report of the donors and amounts donated.
+
+    :returns: the donor report as a string.
+    """
+    # First, reduce the raw data into a summary list view
+    report_rows = []
+    for (name, gifts) in donor_db.values():
+        total_gifts = sum(gifts)
+        num_gifts = len(gifts)
+        avg_gift = total_gifts / num_gifts
+        report_rows.append((name, total_gifts, num_gifts, avg_gift))
+
+    # sort the report data
+    report_rows.sort(key=sort_key)
+    report = []
+    report.append("{:25s} | {:11s} | {:9s} | {:12s}".format("Donor Name",
+                                                            "Total Given",
+                                                            "Num Gifts",
+                                                            "Average Gift"))
+    report.append("-" * 66)
+    for row in report_rows:
+        report.append("{:25s}   ${:10.2f}   {:9d}   ${:11.2f}".format(*row))
+    return "\n".join(report)
+
+
+def save_letters_to_disk():
+    """
+    make a letter for each donor, and save it to disk.
+    """
+    for donor in donor_db.values():
+        letter = gen_letter(donor)
+        # I don't like spaces in filenames...
+        # create data_out directory to hold generated letters
+        filename = "data_out/" + donor[0].replace(" ", "_") + ".txt"
+        os.makedirs(os.path.dirname(filename), exist_ok=True)
+        print("writting letter to:", donor[0])
+        with open(filename, 'w') as f:
+            f.write(letter)
+
+
+def print_donor_report():
+    print(generate_donor_report())
+
+
+def quit():
+    """
+    quit the program
+
+    Note: you could put the sys.exit call directly in the dict
+          but this lets you put extra code in there if you want.
+    """
+    sys.exit(0)
+
+
+if __name__ == "__main__":
+
+    donor_db = get_donor_db()
+
+
+    selection_dict = {"1": send_thank_you,
+                      "2": print_donor_report,
+                      "3": save_letters_to_disk,
+                      "4": quit}
+
+    while True:
+        selection = main_menu_selection()
+        try:
+            selection_dict[selection]()
+        except KeyError:
+            print("\nError: menu selection is invalid!")
```

Comment:
all looking good -- but do make sure you "get" it all -- particularly the dispatch dict for the main menu!